be/src/util/file_cache.cpp

```diff
@@ -0,0 +1,416 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "util/file_cache.h"
+
+#include <atomic>
+#include <cstdint>
+#include <cstring>
+#include <memory>
+#include <mutex>
+#include <ostream>
+#include <string>
+#include <utility>
+#include <vector>
+#include <chrono>
+#include <functional>
+
+#include "env/env.h"
+#include "gutil/strings/substitute.h"
+
+namespace doris {
+
+namespace internal {
+
+template <class FileType>
+class OpenedFileHandle;
+
+// Encapsulates common descriptor fields and methods.
+template <class FileType>
+class BaseDescriptor {
+public:
+    BaseDescriptor(FileCache<FileType>* file_cache, const std::string& filename)
+        : _file_cache(file_cache), _file_name(filename) { }
+
+    ~BaseDescriptor() {
+        // The (now expired) weak_ptr remains in '_descriptors', to be removed
+        // by the next call to RunDescriptorExpiry(). Removing it here would
+        // risk a deadlock on recursive acquisition of '_lock'.
+
+        if (deleted()) {
+            cache()->erase(filename());
+
+            WARN_IF_ERROR(env()->delete_file(filename()), "delete file failed:");
+        }
+    }
+
+    // Insert a pointer to an open file object(FileType*) into the file cache with the
+    // filename as the cache key.
+    //
+    // Returns a handle to the inserted entry. The handle always contains an
+    // open file.
+    OpenedFileHandle<FileType> insert_into_cache(void* file_ptr) const {
+        // TODO(hkp)
+        auto deleter = [](const doris::CacheKey& key, void* value) {
+            delete (FileType*)value;
+        };
+        FileType* file = reinterpret_cast<FileType*>(file_ptr);
+        CacheKey key(file->file_name());
+        auto lru_handle = cache()->insert(key, file_ptr, sizeof(file_ptr), deleter);
```

Comment:
The third argument should be 1, since cache's capacity is set to max_open_files.

Reply:
done